### PR TITLE
Align KTO with DPO: Align compute_loss flow

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Align compute_loss flow.

This PR aligns KTOTrainer's metric storage and compute-loss architecture with DPOTrainer: to simplify loss computation and metrics handling, and to streamline evaluation logic. The main changes focus on consolidating loss and metrics calculation, updating how metrics are stored, and simplifying the prediction step.

Part of:
- #4786

### Motivation

DPOTrainer appends metrics directly to self._metrics[mode] inside _compute_loss / _compute_loss_liger, with no intermediate dict and no helper method.

KTOTrainer used a different pattern (get_batch_loss_metrics → store_metrics) that also included an is_main_process guard absent from DPO. The is_main_process guard was unnecessary: gather_for_metrics already produces the same aggregated scalar on every rank, so appending unconditionally (as DPO does) gives identical results.

### Solution

- Rename get_batch_loss_metrics → _compute_loss(model, inputs, return_outputs) to match DPO's naming convention.
- Move metric appending inside _compute_loss: instead of building a metrics dict and returning it to the caller, append directly to self._metrics[mode], exactly as DPO does in its _compute_loss.
- Simplify compute_loss to a one-liner that delegates to _compute_loss.
- Simplify prediction_step to call compute_loss and return (loss, None, None), matching DPO's pattern (the previous implementation returned fake logits derived from intermediate metric sums).
- Remove store_metrics (no longer needed) and its now-unused Literal import.

### Changes

**Refactoring and Simplification:**

* The methods `get_batch_loss_metrics` and `compute_loss` have been merged and refactored into `_compute_loss` and a new, simplified `compute_loss` method, reducing code duplication and clarifying responsibilities.
* The `prediction_step` method is rewritten to always use `compute_loss`, ensuring consistent loss computation for both training and evaluation, and removing custom metrics/logit handling in favor of a simpler interface.

**Metrics Handling Improvements:**

* Metrics are now directly appended to the internal `_metrics` dictionary within `_compute_loss`, removing the need for a separate `store_metrics` method and streamlining metrics collection.

**Type and Import Cleanups:**

* The unused `Literal` import was removed from the type imports, reflecting updates to function signatures.

---

> [!NOTE]
> **Medium Risk**
> Touches core `Trainer` integration points (`compute_loss`/`prediction_step`) and changes how/when metrics are recorded during train/eval, which can affect evaluation outputs and logging behavior.
> 
> **Overview**
> Refactors `KTOTrainer` to compute loss via a single `_compute_loss` path and **append metrics directly into `self._metrics[train|eval]`**, removing the prior `get_batch_loss_metrics` + `store_metrics` pattern and associated main-process guard.
> 
> Simplifies `compute_loss` to delegate to `_compute_loss`, and rewrites `prediction_step` to always run loss computation during eval (returning dummy logits/labels when needed so `compute_metrics` still runs). Also removes the now-unused `Literal` import and tightens some method signatures.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d8264c7df5cd78525dc6c24c2ba0d2151e64a44d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -1112,63 +1113,43 @@ def get_batch_loss_metrics(
...
             )
-            metrics["count/rejected"] = all_num_rejected
+            self._metrics[mode]["count/rejected"].append(all_num_rejected)
 
         loss = losses.nanmean()
         if self.aux_loss_enabled:
             loss += self.aux_loss_coef * aux_loss
 
-        return loss, metrics
-
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/experimental/kto/kto_trainer.py:1149]: good point, either fix it here or in a future pr

**Review by qgallouedec — APPROVED** [review]

File: trl/experimental/kto/kto_trainer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (albertvillanova) [reply trl/experimental/kto/kto_trainer.py:1149]: Noted for a following PR.
